Add input validation to entry widget

Jacalz opened the pull request:
### Description:
This PR adds the initial steps towards input validation within fyne. It adds the necessary support for entry widgets which should later be possible to expand to form and soft keyboards as proposed in #916.

Partial solution for #916 

The implementation also works with password entries as you can see:
![valid-password](https://user-images.githubusercontent.com/25466657/90391520-90a58a80-e08d-11ea-8354-edb9863ce982.gif)

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [x] Public APIs match existing style.
- [ ] Any breaking changes have a deprecation path or have been discussed.

charlesdaniels commented:
I like this approach.

However I think it fails to give the user any information about whether or not the input was accepted. 

This is my suggestion:

![IMG_0227](https://user-images.githubusercontent.com/11738895/85415496-6034ea80-b53b-11ea-9881-169cc89b1063.PNG)

* The colors could be based on the theme. I haven't dug into the theming system much, but I think something like a `ThemeColorAccentPositiveConnotation()` and `ThemeColorAccentNegativeConnotation()` would make sense, if we don't have something equivalent already.

* I think the icons are important, rather than just changing color, since colorblind users may not be able to distinguish the colors.

* There should be some way to convey to the user what a valid input is. I don't like tooltips for this, since they aren't discoverable.

* If there isn't already an appropriate variant of `Entry`, having a generic `EntryWithIcon` would be a good place to cut the abstraction for this, as other things might want that.

* Note for clarity: I would not modify non-validate entries in appearance from how the look now, since the sketch above might insinuate the contrapositive.

Jacalz commented:
Just rebased on develop and added an initial version of a little icon for displaying if the input is valid or not:
![validation-preview](https://user-images.githubusercontent.com/25466657/90336853-d0616900-dfde-11ea-8545-10199ed1efdf.gif)

andydotxyz commented on cmd/fyne_demo/screens/widget.go:
```diff
@@ -174,6 +175,11 @@ func makeInputTab() fyne.Widget {
 	entryDisabled.Disable()
+	entryValidated := widget.NewValidatedEntry(regexp.MustCompile(`\d`))
+	entryValidated.SetPlaceHolder("Validated, must contain a number")
+	entryValidated.OnChanged = func(validText string) {
```
Is this correct? Is OnChanged only called when the input has been validated now?

Jacalz replied:
Yeah that was the initial way that I had this set up. I think it can be quite neat to only have the callback run when it is valid. Gets rid of a lot of manual code for checking if we are correct. Do you prefer to always run it, even with invalid input?

andydotxyz commented on widget/entry.go:
```diff
@@ -41,6 +42,10 @@ type Entry struct {
 	MultiLine   bool
 	Wrapping    fyne.TextWrap
 
+	RegexValidation *regexp.Regexp
```
I just thought - RegexValidation also implies there could be multiple types of validation (which I agree with) - but why bake this specific one into entry? Perhaps "Validation" can be provided by a "RegexpValidation" provider?

andydotxyz commented:
A few comments inline. Mostly the tying to Regex(p) concerns me.
Also, can we follow the design guide and use a red/error colour for invalid state?

Jacalz commented on widget/entry.go:
```diff
@@ -41,6 +42,10 @@ type Entry struct {
 
+	RegexValidation *regexp.Regexp
+	regexStatus     *regexStatus
+	ValidInput      bool // Will be false if RegexValidation is empty
```
Yeah good call. Shouldn't really be modified by the user. Better to have a method for accessing it.

Jacalz commented:
I don't know about having a constructor. The multiline entry is also just setting one value so that falls into the same category somewhat though ;) But yeah, it definitely isn't a necessity to have. I just need to make sure that tests work without it :)

Jacalz commented on theme/icons.go:
```diff
@@ -78,6 +83,37 @@ func (res *InvertedThemedResource) Invert() *ThemedResource {
+	return colorizeResource(res.source, clr)
+}
+
+// // Warning adapts the color of the resource to the warning color.
```
It isn't there. You were checking through an old commit ;)

Jacalz commented on theme.go:
```diff
@@ -21,6 +21,7 @@ type Theme interface {
 	FocusColor() color.Color
 	ScrollBarColor() color.Color
 	ShadowColor() color.Color
+	// ErrorColor() color.Color
```
It could have been used as a reminder for 2.0 when we can add it in, but I think we can keep track of that anyway. Better remove it then :)

Jacalz commented on theme/theme.go:
```diff
@@ -54,6 +55,7 @@ func DarkTheme() fyne.Theme {
 		hover:          color.NRGBA{0xff, 0xff, 0xff, 0x0f},
 		scrollBar:      color.NRGBA{0x0, 0x0, 0x0, 0x99},
 		shadow:         color.NRGBA{0x0, 0x0, 0x0, 0x66},
+		errors:         color.NRGBA{0xf4, 0x43, 0x36, 0xff},
```
Indeed. Better to add it all in with theme changes later down the road to 2.0 :)

Jacalz commented:
The implementation here should now be pretty much ready for prime-time except for the missing helper text for explaining why validation failed.

andydotxyz commented:
Looking good.
I think Reason might want a fallback text, but until we display it that really doesn't matter :)

Jacalz commented:
Pushed some changes now per the discussions in Slack. There is now an `entry.SetValidationError(error)` that sets a validation manually, wording has moved from "Regex" to "Regexp", internal status is now of error type and there are a few smaller test, formatting and linter changes.

stuartmscott commented on widget/entry_validation.go:
```diff
@@ -0,0 +1,85 @@
+	}
+}
+
+var erroricon = theme.NewErrorThemedResource(theme.ErrorIcon()) // Avoid having to parse XML on each status refresh
```
Note: this will not update if the theme changes

andydotxyz replied:
Moving this to the CreateRenderer should not be too slow as the XML parse and render phase result is cached using the logic in internal/painter/svg_cache.go.

Jacalz replied:
I will just remove this step entirely and instead quote the wise @andydotxyz, by saying "Best to leave optimisations until we see that they are needed" :wink:

stuartmscott commented on validator.go:
```diff
@@ -0,0 +1,27 @@
+	Reason string
+	Regexp *regexp.Regexp
+}
+
```
Consider adding `func NewRegexpValidator(regexp, reason string) *RegexpValidator` that can wrap the `regexp.MustCompile`

andydotxyz replied:
Agreed - except I would not use MustCompile in that case as it will cause panic if errored.
Instead try the compile and log an error if not successful I think? (and don't perform any validation) - better than an app dying :)

andydotxyz commented:
Agreed with Stuart's comments except minor warning about MustCompile.
Otherwise I'm happy once his comments are addressed.

Jacalz commented:
Alright. Should be all set now hopefully :slightly_smiling_face:

stuartmscott commented on theme/icons.go:
```diff
@@ -78,6 +83,29 @@ func (res *InvertedThemedResource) Invert() *ThemedResource {
+
+// Content returns the underlying content of the resource adapted to the current background color.
+func (res *ErrorThemedResource) Content() []byte {
+	clr := &color.NRGBA{0xf4, 0x43, 0x36, 0xff} // Should be current().ErrorColor() in the future
```
nit: it is easy to forget this comment. Consider adding a "TODO" or "FIXME" prefix so it shows up in searches

stuartmscott commented on cmd/fyne_demo/screens/widget.go:
```diff
@@ -175,6 +175,8 @@ func makeInputTab() fyne.Widget {
 	entryDisabled := widget.NewEntry()
 	entryDisabled.SetText("Entry (disabled)")
 	entryDisabled.Disable()
+	entryValidated := &widget.Entry{Validator: fyne.NewRegexpValidator(`\d`, "Does not contain a number")}
```
nit:
```suggestion
	entryValidated := &widget.Entry{Validator: fyne.NewRegexpValidator(`\d`, "Must contain a number")}
```

stuartmscott commented on validator_test.go:
```diff
@@ -0,0 +1,20 @@
+
+	r.Regexp = regexp.MustCompile(`^\d{2}-\w{4}$`)
+
+	assert.Equal(t, nil, r.Validate("10-fyne"))
+	assert.Equal(t, errors.New(r.Reason), r.Validate("100-fynedesk"))
```
The assert library has special assertions for errors
```suggestion
	assert.NoError(t, r.Validate("10-fyne"))
	assert.EqualError(t, r.Validate("100-fynedesk"), r.Reason)
```

stuartmscott commented on widget/entry.go:
```diff
@@ -1083,6 +1112,13 @@ func (r *entryRenderer) Refresh() {
 	}
+	if r.entry.Validator != nil {
+		if !r.entry.Focused() && r.entry.Text != "" && r.entry.validationError != nil {
+			r.line.FillColor = &color.NRGBA{0xf4, 0x43, 0x36, 0xff}
```
nit: add a similar TODO comment here so the hardcoded colour is replaced in future

stuartmscott commented on widget/entry_test.go:
```diff
@@ -43,6 +43,18 @@ func TestEntry_passwordRevealerCursor(t *testing.T) {
+	assert.Error(t, r.Validate(entry.Text))
+
+	test.Type(entry, "-12")
+	assert.Equal(t, nil, r.Validate(entry.Text))
```
```suggestion
	assert.NoEqual(t, r.Validate(entry.Text))
```

Jacalz replied:
I think you meant assert.NoError ;)

Jacalz commented:
Thanks for those small suggestions @stuartmscott. Should be fixed now :)
